apache/superset
# feat(formatters): Add custom d3-time-format locale

### SUMMARY

Adds a configuration option to change the default language of time related data on charts, including the option to translate months. This is related to issues: #3972 , https://github.com/apache/superset/issues/13442 and https://github.com/apache/superset/issues/17447.

During the implementation I had to change how the smart formatters were used across the application, because I had to ensure that they were instantiated with the right locale configuration. So instead of importing them directly, I got them through the TimeFormatterRegistry. I'm not sure if is the best option, let me know.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

Without time format configuration:
![Captura de tela de 2023-06-01 09-49-00](https://github.com/apache/superset/assets/14370340/facbbc2d-04b2-47d4-8240-472c85455bff)

With brazilian portuguese time format configuration:
![Captura de tela de 2023-06-01 09-47-58](https://github.com/apache/superset/assets/14370340/e627ab9e-444c-4f32-bdee-34b32348c38b)



### TESTING INSTRUCTIONS

Add the `D3_TIME_FORMAT` with your locale configuration to `superset_config.py` following the definition expected by D3 ([docs](https://github.com/d3/d3-time-format/tree/main#locales)), for example, this is the definition for brazilian portuguese:

```py
D3_TIME_FORMAT = {
  "dateTime": "%A, %e de %B de %Y. %X",
  "date": "%d/%m/%Y",
  "time": "%H:%M:%S",
  "periods": ["AM", "PM"],
  "days": ["Domingo", "Segunda", "Terça", "Quarta", "Quinta", "Sexta", "Sábado"],
  "shortDays": ["Dom", "Seg", "Ter", "Qua", "Qui", "Sex", "Sáb"],
  "months": ["Janeiro", "Fevereiro", "Março", "Abril", "Maio", "Junho", "Julho", "Agosto", "Setembro", "Outubro", "Novembro", "Dezembro"],
  "shortMonths": ["Jan", "Fev", "Mar", "Abr", "Mai", "Jun", "Jul", "Ago", "Set", "Out", "Nov", "Dez"]
}
```


### ADDITIONAL INFORMATION
